pkg/tfcli/client.go

```diff
@@ -0,0 +1,321 @@
+		logger.Info("Waiting for process to terminate gracefully.", "arg-from-process", c.pInfo.GetCmd().String())
+		err := c.pInfo.WaitError()
+		stderr, _ := c.pInfo.StderrAsString()
+		stdout, _ := c.pInfo.StdoutAsString()
+		logger = logger.WithValues("stderr", stderr, "stdout", stdout)
+
+		if err != nil {
+			logger.Info("Failed to run Terraform CLI", "error", err)
+		} else {
+			logger.Info("Successfully executed Terraform CLI", "stdout", stdout)
+		}
+		if storeResult != nil {
+			storePath := filepath.Join(c.wsPath, fileStore)
+			if err := storeResult(c, stdout, stderr); err != nil {
+				logger.Info("Failed to store result from Terraform CLI", "store", storePath)
+				return
+			}
+			logger.Info("Successfully stored result from Terraform CLI", "store", storePath)
+		}
+	}()
+	return nil
+}
+
+func (c *client) syncPipeline(ctx context.Context, ignoreExitErr bool, command string, args ...string) error {
+	var err error
+	if c.pInfo, err = process.New(command, args, c.wsPath, false, true, false, c.logger.log); err != nil {
+		return errors.Wrapf(err, fmtErrSyncRun, command, args, c.wsPath)
+	}
+	exitErr := &exec.ExitError{}
+	if err = c.pInfo.Run(ctx); err != nil && (!ignoreExitErr || !errors.As(err, &exitErr)) {
+		return errors.Wrapf(err, fmtErrSyncRun, command, args, c.wsPath)
+	}
+	return nil
+}
+
+// returns true if state file exists
+// TODO(aru): differentiate not-found error
+func (c *client) loadStateFromWorkspace(errNoState bool) (bool, error) {
+	pathState := filepath.Join(c.wsPath, fileState)
+	var err error
+	c.state.tfState, err = ioutil.ReadFile(pathState)
+	if err == nil {
+		return true, nil
+	}
+	if !os.IsNotExist(err) || errNoState {
+		return false, errors.Wrapf(err, fmtErrLoadState, pathState)
+	}
+	// then state not found
+	return false, nil
+}
+
+func (c *client) storeStateInWorkspace() error {
+	if c.state.tfState == nil {
+		return nil
+	}
+
+	pathState := filepath.Join(c.wsPath, fileState)
+	return errors.Wrapf(ioutil.WriteFile(pathState, c.state.tfState, 0600), fmtErrStoreState, pathState)
+}
+
+func pathExists(path string, isDir bool) (bool, error) {
+	fInfo, err := os.Stat(path)
+	if err == nil && fInfo.IsDir() == isDir {
+		// path exists and is of expected type
+		return true, nil
+	}
+	if err == nil && fInfo.IsDir() != isDir {
+		return false, errors.Errorf(fmtErrPath, path, isDir, fInfo.IsDir())
+	}
+	if err != nil && !os.IsNotExist(err) {
+		return false, errors.Wrapf(err, fmtErrPath, path, isDir, fInfo.IsDir())
+	}
+	return false, nil
+}
+
+func (c *client) closeOnError(ctx context.Context, f func() error) error {
+	err := f()
+	if err != nil {
+		err = multierr.Combine(err, c.Close(ctx))
+	}
+	return err
+}
+
+// Close releases resources allocated for this client.
+// After a call to Close, do not reuse the same handle.
+func (c *client) Close(_ context.Context) error {
+	var result error
+	if c.pInfo != nil {
+		result = multierr.Combine(errors.Wrap(c.pInfo.Kill(), errDestroyProcess))
```

Comment:
`tfcli.Client` can be running a single process at a given time. `pInfo` is related to that process if set. Common controller calls `Close` after an MR is deleted successfully. We don't aim to keep track of all Terraform CLI process forks done in the context of an MR via multiple Clients initialized with the same handle corresponding to the same workspace. The workspace state prevents that scenario anyway. 

Relying on Terraform state locks, however, we do need a _stale_ lock cleanup mechanism as we have been discussing. We will address that in a separate PR.

Reply:
I'm not sure I got the relation with locks. What I meant is that, from my reading, `c.pInfo` will be non-nil only if the operation is started in this reconciliation. In the next one, i.e. in-progress state, we initialize a new client but don't call `process.New` function because a process is in progress. Then when we trigger a deletion we don't close right away, we check its progress in the next reconciliation and close it only if it's completed. But then `c.pInfo` is `nil` because that's a new reconciliation where `c.pInfo` wasn't initialized. So `Close(ctx)` may not be cleaning up anything because in line 258 `p.Info` will be `nil` all the time.

Reply:
For async operations, we do not hold references to forked processes across reconciliation boundaries. Currently, we do not need to that because we are observing the workspace state. We will block on deletion if there is an ongoing command.